### DISCUSSION TITLE
Node and edge(-objects)

### DISCUSSION THREAD

**Discussion by IjonTichyRaumpilot** [networkx/networkx#4881] [Q&A] [answered]
I am having problems in using nodes and edges. 
So what I want is, when I create a node, I want a node object where all attributes are attched. So that is possible and works as expected. So I write g.add_node(node). And a node can have any attribute like, is_XY etc. So when I iterate over all nodes by e.g. for n in G.nodes() I run over all nodes and can check n.isXY etc. Nice and easy to understand.

Now, for edges I want to do the same, but that seems to be more complex then needed.
When I create an edge e.g. g.add_edge(source,target) I also want that edge to be an object where I can add attribute to. So that I can easily iterate over all edges and checking things. e.g. for e in G.edges(): That seems to be a bit complex. I can set an object as an attribute by using add_edge(source,target,edge=object), but then accessing it is a bit unintuitive. I found I would need to use G.edges()[e]['edge'].is_XY. 
Is there an easier/more intuitive way to do this. I would assume that we need to have a class that implements an abstract edge class (sorry did some java before coming to python).
Any hints are appreciated.

Here the example

`import networkx as nx

class node:
    name=""
    is_XY=False
    def __init__(self,name,is_XY):
        self.name=name
        self.is_XY=is_XY
    
G = nx.DiGraph()
n1=node("n1",True)
G.add_node(n1)
n2=node("n2",False)
G.add_node(n2)
n3=node("n3",False)
G.add_node(n3)

for n in G.nodes():
    if n.is_XY:
        print('Node with is_XY==True: '+n.name)

class edge:
    name=""
    is_XY=False
    def __init__(self,name,is_XY):
        self.name=name
        self.is_XY=is_XY 

e1=edge("e1",True)
G.add_edge(n1,n2,edge=e1)
e2=edge("e2",False)
G.add_edge(n1,n3,edge=e2)

for e in G.edges():
    if G.edges()[e]['edge'].is_XY:
        #I would love if something like e.is_XY instead
        print('Edge with is_XY==True: '+G.edges()[e]['edge'].name)`

**Comment by MridulS** [comment, marked as answer]
Is having an object with attributes absolutely required? Something like this can be done via node and edge attributes too.
``` python

In [1]: import networkx as nx

In [2]: G = nx.Graph()

In [3]: G.add_node("n1", is_XY=True)

In [4]: G.add_node("n2", is_XY=False)

In [5]: G.add_node("n3", is_XY=False)

In [6]: G.add_edge("n1", "n2", is_XY=True)

In [7]: G.add_edge("n2", "n3", is_XY=False)

In [8]: for n, attr in G.nodes.data():
    ...:     if attr['is_XY']:
    ...:         print(f"Node with is_XY==True:", n)
    ...:
    ...:
             Node with is_XY==True: n1

In [9]: for source, target, attr in G.edges.data():
    ...:     if attr['is_XY']:
    ...:         print(f"Edge with is_XY==True:", source, target)
    ...:
           Edge with is_XY==True: n1 n2
```

**Reply by IjonTichyRaumpilot** [reply]
Thanks for the answer. I think that's the way to go.
